File: backend/src/main/java/es/udc/tfgproject/backend/model/services/ShoppingService.java
package es.udc.tfgproject.backend.model.services;

import java.math.BigDecimal;

import es.udc.tfgproject.backend.model.entities.DiscountTicket;
import es.udc.tfgproject.backend.model.entities.Order;
import es.udc.tfgproject.backend.model.entities.ShoppingCart;
import es.udc.tfgproject.backend.model.exceptions.CompanyDoesNotAllowHomeSaleException;
import es.udc.tfgproject.backend.model.exceptions.DiscountTicketHasExpiredException;
import es.udc.tfgproject.backend.model.exceptions.DiscountTicketUsedException;
import es.udc.tfgproject.backend.model.exceptions.EmptyShoppingCartException;
import es.udc.tfgproject.backend.model.exceptions.IncorrectDiscountCodeException;
import es.udc.tfgproject.backend.model.exceptions.InstanceNotFoundException;
import es.udc.tfgproject.backend.model.exceptions.PermissionException;

public interface ShoppingService {

	ShoppingCart addToShoppingCart(Long userId, Long shoppingCartId, Long productId, Long companyId, int quantity)
			throws InstanceNotFoundException, PermissionException;

	ShoppingCart updateShoppingCartItemQuantity(Long userId, Long shoppingCartId, Long productId, Long companyId,
			int quantity) throws InstanceNotFoundException, PermissionException;

	ShoppingCart removeShoppingCartItem(Long userId, Long shoppingCartId, Long productId, Long companyId)
			throws InstanceNotFoundException, PermissionException;

	ShoppingCart findShoppingCartProducts(Long userId, Long shoppingCartId, Long companyId)
			throws InstanceNotFoundException, PermissionException;

	ShoppingCart changeShoppingCartHomeSale(Long userId, Long shoppingCartId, Long companyId, Boolean homeSale)
			throws InstanceNotFoundException, PermissionException, CompanyDoesNotAllowHomeSaleException;

	Order buy(Long userId, Long shoppingCartId, Long companyId, Boolean homeSale, String street, String cp,
			String codeDiscount) throws InstanceNotFoundException, PermissionException, EmptyShoppingCartException,
			IncorrectDiscountCodeException, DiscountTicketHasExpiredException, DiscountTicketUsedException;

	Order findOrder(Long userId, Long orderId) throws InstanceNotFoundException, PermissionException;

	Block<Order> findUserOrders(Long userId, int page, int size);

	Block<Order> findCompanyOrders(Long userId, Long companyId, int page, int size)
			throws InstanceNotFoundException, PermissionException;

	BigDecimal redeemDiscountTicket(Long userId, Long companyId, Long shoppingCartId, String code)
			throws InstanceNotFoundException, PermissionException, IncorrectDiscountCodeException,
			EmptyShoppingCartException, DiscountTicketHasExpiredException, DiscountTicketUsedException;

	Block<DiscountTicket> findUserDiscountTicketsNotUsed(Long userId, int page, int size)
			throws InstanceNotFoundException;

}

File: frontend/src/backend/shoppingService.js
import {config, appFetch} from './appFetch';

export const addToShoppingCart = (shoppingCartId, productId, companyId, quantity, onSuccess, 
    onErrors) =>
    appFetch(`/shopping/shoppingCarts/${shoppingCartId}/addToShoppingCart`, 
        config('POST', {productId, companyId, quantity}), onSuccess, onErrors);

export const updateShoppingCartItemQuantity = (shoppingCartId, productId, companyId,
    quantity, onSuccess, onErrors) =>
    appFetch(`/shopping/shoppingCarts/${shoppingCartId}/updateShoppingCartItemQuantity`, 
        config('POST', {productId, companyId, quantity}), onSuccess, onErrors);

export const removeShoppingCartItem = (shoppingCartId, productId, companyId, onSuccess,
    onErrors) =>
    appFetch(`/shopping/shoppingCarts/${shoppingCartId}/removeShoppingCartItem`, 
        config('POST', {productId, companyId}), onSuccess, onErrors);

export const findShoppingCartProducts = (shoppingCartId, companyId,  onSuccess) => 
    appFetch(`/shopping/shoppingCarts/${shoppingCartId}?companyId=${companyId}`, 
        config('GET'), onSuccess);

export const changeShoppingCartHomeSale = (shoppingCartId, companyId, homeSale,
    onSuccess, onErrors) =>
    appFetch(`/shopping/shoppingCarts/${shoppingCartId}/changeShoppingCartHomeSale`, 
        config('POST', {companyId, homeSale}), onSuccess, onErrors);
        

export const buy = (shoppingCartId, companyId, homeSale, street, cp,
    onSuccess, onErrors) =>
    appFetch(`/shopping/shoppingCarts/${shoppingCartId}/buy`, 
        config('POST', {companyId, homeSale, street, cp}), onSuccess, onErrors);

export const findUserOrders = ({page}, onSuccess) => 
    appFetch(`/shopping/userOrders?page=${page}`, config('GET'), onSuccess);

export const findCompanyOrders = (companyId, {page}, onSuccess) => 
    appFetch(`/shopping/companyOrders?companyId=${companyId}&page=${page}`, config('GET'), onSuccess);

export const findOrder = (orderId, onSuccess) =>
    appFetch(`/shopping/orders/${orderId}`, config('GET'), onSuccess);
        